lib/Frontend/Frontend.cpp

```diff
@@ -938,11 +938,13 @@ CompilerInstance::getInputBuffersIfPresent(const InputFile &input) {
   // FIXME: Working with filenames is fragile, maybe use the real path
   // or have some kind of FileManager.
   using FileOrError = llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>;
+  // Avoid memory-mapping when the compiler is run for IDE inspection,
+  // since that would prevent the user from saving the file.
   FileOrError inputFileOrErr =
     swift::vfs::getFileOrSTDIN(getFileSystem(), input.getFileName(),
                               /*FileSize*/-1,
                               /*RequiresNullTerminator*/true,
-                              /*IsVolatile*/false,
+                              /*IsVolatile*/getInvocation().isIDEInspection(),
```

Comment:
A few thoughts, unfortunately without a real suggestion.

`isIDEInspection()` will be false if we are not in a completion-like operator (like getting diagnostics or the old cursor info implementation). In those cases we do go through `SwiftASTManager`, however, so maybe we don’t hit this function and we only run the `getBufferForFile` that you also changed.

The real question here, I think, is whether we are running inside of `sourcekitd` or the compiler. But I don’t have any idea of how to check that.

@rintaro Do you have any ideas/thoughts? About my comment and in general with this PR.

Reply:
I'm happy to use a more general condition of we can come up with one. Otherwise, it seems like the one I'm using is a valid improvement, which I would like to reap the developer experience benefits of since this issue is painful on a daily basis for Browser Company engineers.

Reply:
I'm happy to take this for now 👍